### PR TITLE
check if the receiver parameter of @MustCallChoice is a resouce varia…

### PR DESCRIPTION
**Pull request by Nargeshdb**

### CODE REVIEW

**Comment by msridhar** [comment]
Rather than having to sprinkle these checks in everywhere, I wonder if it would be cleaner to have the Must Call Checker remove `close` from the `@MustCall` type of any try-with-resources variable.  @kelloggm do you think that would be doable?

**Comment by mernst** [comment]
> Rather than having to sprinkle these checks in everywhere, I wonder if it would be cleaner to have the Must Call Checker remove `close` from the `@MustCall` type of any try-with-resources variable. @kelloggm do you think that would be doable?

I agree with this proposal.  (That's how I assumed you were planning to handle it.)

**Comment by kelloggm** [comment]
Yes, that's do-able. I think I could implement it that way pretty easily. I'll make a PR when I get a chance.

**Comment by kelloggm** [comment]
@msridhar @Nargeshdb I've implemented what Manu suggested in the MustCall Checker. Please have a look.

The tests are still failing. I think that the MC checker is doing its job here, and there is something wrong with both of the test methods:

1. the first error is caused by the file descriptor problem we discussed earlier. We should probably remove this test from this PR, because it'll need to be addressed separately (e.g. in #262). The error is:
    ```  
    object-construction-checker/tests/socket/TryWithResourcesSimple.java:35: error: [required.method.not.called] @MustCall method(s) close,  for variable/expression not invoked.  The type of object is: java.nio.channels.FileChannel.  Reason for going out of scope: never assigned to an @Owning location
            file.setLength(outSpecial.getChannel().position());
    ```

2. the second error is caused by use not having a stub file for `RandomAccessFile` with an appropriate `@MustCallChoice` annotation:
    ```
    object-construction-checker/tests/socket/TryWithResourcesSimple.java:35: error: [required.method.not.called] @MustCall method(s) close,  for variable/expression not invoked.  The type of object is: java.nio.channels.FileChannel.  Reason for going out of scope: never assigned to an @Owning location
            file.setLength(outSpecial.getChannel().position());
    ```
    A similar test using `Socket` and `getInputStream` works in the Must Call Checker tests, so this should too once we have that stub.

File: must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java
Comment (msridhar) [inline must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java:100]: Is it possible to define a class that implements `Closeable` and also has additional `@MustCall` requirements?  Admittedly a corner case, but if it's possible, we may want to tweak this logic to not set the type to bottom, but instead to just remove the `close` method from the type.  Thoughts?

Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java:100]: Ah - you're right. I'll fix this.

Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java]: @msridhar I've rewritten to handle this case, and I added a stub for one of the failures in the tests. If we remove the other test, we can merge this if you all are okay with it.

File: must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java
Comment (msridhar) [inline must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: @kelloggm do we still need this change?  If we do, we should update the Javadoc as it's out of date.

**Review by msridhar — CHANGES_REQUESTED** [review]
I removed the failing test case.  One more comment

File: must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java
Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: Nice catch - I didn't think about this when I made the other change last night. This is trickier.

The reason we need this at all is that the RHS of assignments in the `try` block still might have "close" in their MC type. For example, in this code:
```
try (Socket s = new Socket(addr, port)) { } catch (Exception e) { }
```
`Socket s` (the LHS) is correctly assigned the type `@MustCall({})` by taking it's normal type and removing "close", because it's a try-with-resources variable. The RHS, however, has no indication that it's a try-with-resources variable; it's such a `new Socket` - which correctly has the type `@MustCall("close")`. Without this change, then we issue an incompatible assignment warning here.

I think what we actually need to check here is that if the LHS is a resource variable, then we should remove `"close"` from the RHS type and then do the check. I'll try to get that working.

Comment (msridhar) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: Thanks for the explanation.  Based on what you wrote, I am not sure if you need to make a further change here, other than updating the comment.  Can you give a test case where the initial assignment for a try-with-resources variable _should_ yield an error?  I would hope these variables must be (effectively) final, so they can't be re-assigned.

Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: > I would hope these variables must be (effectively) final, so they can't be re-assigned.

That's correct - the compiler inserts the `final` keyword IIRC.

> Can you give a test case where the initial assignment for a try-with-resources variable should yield an error?

What about this:
```
@MustCall({"close", "myMethod"})
class Foo extends Socket { void myMethod() {} /* also constructors etc */ }
try (Socket s = new Foo(addr, port)) { } catch (Exception e) { }
```

I'm actually not sure how much we care about a case like this, but I think at this point we should warn, because the LHS isn't `@MustCall("myMethod")`.

I think the fix is actually to do the subtyping check without close on both sides in the resource variable case. I think I can do that.

Comment (msridhar) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: I think it's fine to report an error at the initial assignment in this case for now.  Eventually I think the "right" fix would be to have the inferred type of `s` to be `@MustCall("myMethod")` and then to get rid of any special-casing of try-with-resources variables in `MustCallInvokedChecker`.  Then if `s.myMethod()` is invoked within the try block, the code will type check.

Comment (kelloggm) [reply must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java:87]: The current code won't report an error.

I agree that's what the code *should* look like, and I think right now this will work correctly if the declared type of the resource variable is `Foo` rather than `Socket`.

**Review by msridhar — APPROVED** [review]

**Comment by kelloggm** [comment]
@mernst, if you don't mind taking a look at these changes, that would be helpful. I don't think it needs to block merging, though, if @msridhar needs this merged ASAP.

**Comment by kelloggm** [comment]
Merging because I want to re-run Zookeeper with our latest changes.

**Pull request merged**
